chained/water_impl.go

```diff
@@ -0,0 +1,41 @@
+package chained
+
+import (
+	"context"
+	"net"
+
+	"github.com/getlantern/common/config"
+	"github.com/getlantern/errors"
+	"github.com/getlantern/flashlight/v7/ops"
+	"github.com/refraction-networking/water"
+	_ "github.com/refraction-networking/water/transport/v0"
+)
+
+type waterImpl struct {
+	config         *water.Config
+	raddr          string
+	reportDialCore reportDialCoreFn
+	nopCloser
+}
+
+func newWaterImpl(addr string, pc *config.ProxyConfig, reportDialCore reportDialCoreFn) (*waterImpl, error) {
+	wasm := ptSettingBytes(pc, "water_wasm")
+	return &waterImpl{
+		raddr: addr,
+		config: &water.Config{
+			TransportModuleBin: wasm,
+		},
+		reportDialCore: reportDialCore,
+	}, nil
+}
+
+func (d *waterImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn, error) {
+	dialer, err := water.NewDialerWithContext(ctx, d.config)
+	if err != nil {
+		return nil, errors.New("failed to create dialer: %v", err)
```

Comment:
It would be better to wrap the error instead of creating a new one. `fmt.Errorf("failed to create dialer: %w", err)`